### PR TITLE
feat(atenlib): index_select; trace_only ops

### PR DESCRIPTION
**Pull request by justinchuby**
Implement `aten_index_select`; enable tests for transpose.

This change also added the **`trace_only`** argument to mark functions as trace only to work around functions that cannot be compiled. The logic can then be tested, but the argument should be removed later when onnxscript extends support for the syntax.

### CODE REVIEW

**Comment by justinchuby** [comment]
cc @bowenbao

**Comment by BowenBao** [comment]
> We need to fine a way to call the function that have the attribute dim before the input index.

Sounds like a restriction posed from onnx script function. Is there something fundamentally blocking from mixing the order of inputs and attributes? cc @gramalingam

**Comment by gramalingam** [comment]
> > We need to fine a way to call the function that have the attribute dim before the input index.
> 
> Sounds like a restriction posed from onnx script function. Is there something fundamentally blocking from mixing the order of inputs and attributes? cc @gramalingam

No, I don't think there is anything fundamentally blocking it. But we need to add the logic to onnxscript translator to process the parameters accordingly, so it is a missing feature.

File: onnxscript/function_libs/torch_aten/ops/core.py [hunk 1]
```diff
@@ -2174,10 +2176,19 @@ def aten_index_reduce(
     raise NotImplementedError()
 
 
-def aten_index_select(self: TensorType, dim: int, index: TensorType) -> TensorType:
+@torch_op("aten::index_select")
+def aten_index_select(self: TTensor, dim: int, index: TInt) -> TTensor:
     # index_select(Tensor self, int dim, Tensor index) -> Tensor
 
-    raise NotImplementedError()
+    if op.Size(op.Shape(index)) == 0:
```
Comment (gramalingam) [inline onnxscript/function_libs/torch_aten/ops/core.py:2183]: Otherwise, is `index` expected to be 1-dimensional? Or, can it have any other rank?

Comment (justinchuby) [reply onnxscript/function_libs/torch_aten/ops/core.py:2183]: Yes, it should be 1d I think.

Comment (justinchuby) [reply onnxscript/function_libs/torch_aten/ops/core.py:2183]: should I do `op.Reshape(index, [-1])` instead?

Comment (gramalingam) [reply onnxscript/function_libs/torch_aten/ops/core.py:2183]: Yes, if we expect it to be 0d or 1d, I think reshape, as you mention, is a better option.

Comment (justinchuby) [reply onnxscript/function_libs/torch_aten/ops/core.py]: Done. thanks!

**Review by fatcat-z — APPROVED** [review]
Approved.

**Pull request merged**